lib/fileset/default.nix

```diff
@@ -0,0 +1,580 @@
+# TODO: Derive property tests from https://en.wikipedia.org/wiki/Algebra_of_sets
+{ lib }:
+let
+
+  inherit (builtins)
+    isPath
+    isString
+    typeOf
+    pathExists
+    readDir
+    isAttrs
+    ;
+
+  inherit (lib.trivial)
+    min
+    ;
+
+  inherit (lib.attrsets)
+    mapAttrs
+    attrByPath
+    setAttrByPath
+    attrNames
+    filterAttrs
+    ;
+
+  inherit (lib.sources)
+    pathType
+    ;
+
+  inherit (lib.lists)
+    length
+    head
+    tail
+    elem
+    elemAt
+    take
+    commonPrefixLength
+    removePrefix
+    hasPrefix
+    foldl'
+    ;
+
+  inherit (lib.path.parts)
+    deconstruct
+    construct
+    ;
+
+  inherit (lib.strings)
+    splitString
+    substring;
+
+  # Internal file set structure:
+  #
+  # # A set of files
+  # <fileset> = {
+  #   _type = "fileset";
+  #
+  #   # TODO: Indicate that these properties are internal
+  #
+  #   # A tree representation of all included files under the subpath
+  #   tree = <entry>;
+  # };
+  #
+  # # A directory entry value
+  # <entry> =
+  #   # A nested directory
+  #   <directory>
+  #
+  #   # A nested file
+  #   | <file>
+  #
+  #   # A removed file or directory
+  #   # This is represented like this instead of removing the entry from the attribute set because:
+  #   # - It improves laziness
+  #   # - It allows keeping the attribute set as a `builtins.readDir` cache
+  #   | null
+  #
+  # # A directory
+  # <directory> =
+  #   # The inclusion state for every directory entry
+  #   { <name> = <entry>; }
```

Comment:
When this constructor is passed a single item, it does not add a file to the fileset, so it allows multiple representations of the same set of files.
Some ideas; either
 - _always_ use a smart constructor to prevent this from happening (including operations such as intersection), or
 - change the definition of this data structure to be more than a set, or
 - make sure that the difference is never observable, burdening all the functions that read from this, assuming users don't read the structure themselves

In the latter case, you'd want to return an object with methods only, at least to users, which is a bit inefficient, but perhaps quite nice to use. It also adds a bunch more code, so I don't know whether it'd be a good idea to implement that at this point.